apis/placement/v1alpha1/disruptionbudget_types.go

```diff
@@ -0,0 +1,166 @@
+	//
+	// This field is mutually exclusive with the MinAvailable field in the spec; exactly one
+	// of them can be set at a time.
+	//
+	// Defaults to 25%.
+	// +kubebuilder:default="25%"
+	// +kubebuilder:validation:XIntOrString
+	// +kubebuilder:validation:Pattern="^((100|[0-9]{1,2})%|[0-9]+)$"
+	// +optional
+	MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty"`
+
+	// MinAvailable is the minimum number of placements that must be available at any time
+	// despite voluntary disruptions. For example, a setting of 10 would imply that
+	// a voluntary disruption (e.g., an eviction) can only happen if there are at least 11
+	// placements from the linked ClusterResourcePlacement object are applied and available.
+	//
+	// This can be either an absolute value (e.g., 1) or a percentage (e.g., 10%).
+	//
+	// If a percentage is specified, Fleet will calculate the corresponding absolute values
+	// as follows:
+	// * if the linked ClusterResourcePlacement object is of the PickFixed placement type,
+	//   the percentage is against the number of clusters specified in the placement (i.e., the
+	//   length of ClusterNames field in the placement policy);
+	// * if the linked ClusterResourcePlacement object is of the PickAll placement type,
+	//   the percentage is against the total number of clusters being selected by the scheduler
+	//   at the time of the evaluation of the disruption budget;
+	// * if the linked ClusterResourcePlacement object is of the PickN placement type,
+	//   the percentage is against the number of clusters specified in the placement (i.e., the
+	//   value of the NumberOfClusters fields in the placement policy).
+	// The end result will be rounded up to the nearest integer if applicable.
+	//
+	// One may use a value of 0 for this field; in this case, voluntary disruption would be
+	// allowed at any time.
+	//
+	// This field is mutually exclusive with the MaxUnavailable field in the spec; exactly one
+	// of them can be set at a time.
+	//
+	// +kubebuilder:validation:XIntOrString
+	// +kubebuilder:validation:Pattern="^((100|[0-9]{1,2})%|[0-9]+)$"
+	// +optional
+	MinAvailable *intstr.IntOrString `json:"minAvailable,omitempty"`
+}
+
+// ClusterResourcePlacementDisruptionBudgetStatus is the observed state of the
+// ClusterResourcePlacementDisruptionBudget.
+type ClusterResourcePlacementDisruptionBudgetStatus struct {
+	// Number of placement disruptions that are currently allowed.
+	// +optional
+	DisruptionsAllowed int32 `json:"disruptionsAllowed"`
+
+	// Current number of available placements.
+	// +optional
+	CurrentAvailable int32 `json:"currentAvailable"`
+
+	// Minimum desired number of available placements.
+	// +optional
+	DesiredAvailable int32 `json:"desiredAvailable"`
+
+	// Total number of placements counted by this disruption budget.
+	// +optional
+	TotalPlacements int32 `json:"totalPlacements"`
+
+	// TODO: Add a map to track ongoing evictions, to protect against concurrent evictions.
```

Comment:
The PDB status will act as the concurrency control tool in the future, hence decided to add more fields now like

- DisruptionAllowed
- CurrentAvailable
- DesiredAvailable
- TotalPlacements

which will be filled before an eviction is executed

Reply:
if this is only used in the future, why not add them later?  Add it now means we need a PDB controller.

Reply:
Removed PDB status